pkg/ring/lifecycler_test.go

```diff
@@ -246,17 +256,90 @@ func TestCheckReady(t *testing.T) {
 	flagext.DefaultValues(&ringConfig)
 	ringConfig.KVStore.Mock = &MockClient{}
 
+	tokens := Tokens([]uint32{1})
 	r, err := New(ringConfig, "ingester")
 	require.NoError(t, err)
 	defer r.Stop()
-	cfg := testLifecyclerConfig(ringConfig, "ring1")
+	cfg := testLifecyclerConfig(t, ringConfig, "ring1")
 	cfg.MinReadyDuration = 1 * time.Nanosecond
 	l1, err := NewLifecycler(cfg, &nopFlushTransferer{}, "ingester")
-	l1.setTokens([]uint32{1})
+	l1.setTokens(tokens)
 	l1.Start()
 	require.NoError(t, err)
 
 	// Delete the ring key before checking ready
 	err = l1.CheckReady(context.Background())
 	require.Error(t, err)
 }
+
+type noopFlushTransferer struct {
+	lifecycler *Lifecycler
+}
+
+func (f *noopFlushTransferer) StopIncomingRequests()                 {}
+func (f *noopFlushTransferer) Flush()                                {}
+func (f *noopFlushTransferer) TransferOut(ctx context.Context) error { return nil }
+
+func TestTokensOnDisk(t *testing.T) {
+	var ringConfig Config
+	flagext.DefaultValues(&ringConfig)
+	ringConfig.KVStore.Mock = consul.NewInMemoryClient(GetCodec())
+
+	r, err := New(ringConfig, "ingester")
+	require.NoError(t, err)
+	defer r.Stop()
+
+	lifecyclerConfig := testLifecyclerConfig(t, ringConfig, "ing1")
+	lifecyclerConfig.NumTokens = 512
+
+	// Start first ingester.
+	ft := &noopFlushTransferer{}
+	l1, err := NewLifecycler(lifecyclerConfig, ft, "ingester")
+	require.NoError(t, err)
+	l1.Start()
+	// Check this ingester joined, is active, and has 512 token.
+	var expTokens []TokenDesc
+	test.Poll(t, 1000*time.Millisecond, true, func() interface{} {
+		d, err := r.KVClient.Get(context.Background(), ConsulKey)
+		require.NoError(t, err)
+		desc, ok := d.(*Desc)
+		if ok {
+			expTokens = desc.Tokens
+		}
+		return ok &&
+			len(desc.Ingesters) == 1 &&
+			desc.Ingesters["ing1"].State == ACTIVE &&
+			len(desc.Ingesters["ing1"].Tokens) == 0 &&
+			len(desc.Tokens) == 512
+	})
+
+	l1.Shutdown()
+
+	// Start new ingester at same token directory.
+	lifecyclerConfig.ID = "ing2"
+	l2, err := NewLifecycler(lifecyclerConfig, &noopFlushTransferer{}, "ingester")
+	require.NoError(t, err)
+	l2.Start()
+	defer l2.Shutdown()
+
+	// Check this ingester joined, is active, and has 512 token.
+	var actTokens []TokenDesc
+	test.Poll(t, 1000*time.Millisecond, true, func() interface{} {
+		d, err := r.KVClient.Get(context.Background(), ConsulKey)
+		require.NoError(t, err)
+		desc, ok := d.(*Desc)
+		if ok {
+			actTokens = desc.Tokens
+		}
+		return ok &&
+			len(desc.Ingesters) == 1 &&
+			desc.Ingesters["ing2"].State == ACTIVE &&
+			len(desc.Ingesters["ing2"].Tokens) == 0 &&
+			len(desc.Tokens) == 512
+	})
+
+	// Check for same tokens.
+	for i := 0; i < 512; i++ {
```

Comment:
Why don't you replace it with `require.Equal(t, expTokens, actTokens)`?

Reply:
Because it is `ing1` in `expTokens[i].Ingester` and `ing2` in `actTokens[i].Ingester`. So that is never going to be true. Hence checking only the tokens.

Reply:
It would be possible now with normalization turned on.